### PR TITLE
dependency('cuda') support

### PR DESCRIPTION
**Pull request by agurtovoy**
Native support for compiling and linking against the CUDA Toolkit using the `dependency` function: 

```meson
project('CUDA test', 'cpp', meson_version: '>= 0.52.0')
exe = executable('prog', 'prog.cc', dependencies: dependency('cuda'))
```

The `version` and `modules` keywords may be passed to request the use of a specific CUDA Toolkit version and/or additional CUDA libraries, correspondingly:

```meson
dep = dependency('cuda', version : '>=10', modules : ['cublas'])
```

Note that explicitly adding this dependency is only necessary if you are using CUDA Toolkit from a C/C++ file or project, or if you are utilizing additional toolkit libraries that need to be explicitly linked to.

### CODE REVIEW

**Comment by agurtovoy** [comment]
Tested on Windows and Ubuntu.

**Comment by ignatenkobrain** [comment]
Flake8 detected 17 issues on d650b9286df753f3ced1f5c26c04057ab579fba2
Visit https://sider.review/gh/repos/19784232/pulls/6076 to review the issues.

<sub>posted by Sider</sub>

**Comment by obilaniu** [comment]
A much-needed contribution, I hope it is merged in!

Some comments:

- Another commonly-used environment variable that points to a CUDA Toolkit is `CUDA_HOME`.
- It is somewhat important to ensure that the CUDA compiler, `nvcc`, and the `cuda` dependency are coherent. That is why I had imagined that a good way to find the CUDA Toolkit would be to start from `nvcc`'s location. Unfortunately Linux distributions sometimes separate the CUDA Toolkit binaries and libraries and place them within their own hierarchies, and I've been unable to think of a way to handle this.
    - A very ugly alternative would be to modify the compiler search logic to use this module's code for `nvcc` only.
- On yet other distros, the package will install regular `pkg-config` files for `cuda-x.y`, `cudart-x.y`, `cublas-x.y`, `cufft-x.y`, `cusparse-x.y`, `cufftw-x.y`, `cusolver-x.y`, `cuinj64-x.y` and `curand-x.y`, which could be used as well.
- Since `0.52.0` is out already, you might need to bump the declared version numbers to `0.52.1` or `0.53.0`.

In short, the CUDA situation is a mess. Let's hope Meson solves it.

**Comment by agurtovoy** [comment]
Thank you for your comments, @obilaniu!

I did see that `CUDA_HOME` is a thing on Linux, but it appeared to be an unofficial convention, and as such also potentially conflicting with `/usr/local/cuda` symlink (NVIDIA's de-facto standard for selecting the default CUDA version), so I thought it'd be better to collect some real-world use cases on how it's used (and whether it's used often enough) before supporting it. If there is a strong sentiment that it should be supported out-of-the-gate, it'd be easy enough to check `CUDA_HOME` path in addition to `/usr/local/cuda*`. The main question is, when both `/usr/local/cuda` and  `CUDA_HOME` are present and different (and both satisfy the specified version requirements), which one should be considered the default?

Completely agree that mismatching `nvcc` and CUDA Toolit versions should be detected/diagnosed, with the exception of pure C/C++ projects that are linking against the toolkit but not actually compiling any device code. This is already handled by this PR, see https://github.com/mesonbuild/meson/pull/6076/files#diff-ab11a3ea7105654dae69ce1d36a036a2R76.

As far as `pkg-config` files support, I thought this could wait until the next iteration (mainly because I don't have a lot of experience with using CUDA in this configuration).

Good point on the Meson version, will do. Please keep the feedback coming :).

**Comment by obilaniu** [comment]
> I did see that `CUDA_HOME` is a thing on Linux, but it appeared to be an unofficial convention, and as such also potentially conflicting with `/usr/local/cuda` symlink (NVIDIA's de-facto standard for selecting the default CUDA version), so I thought it'd be better to collect some real-world use cases on how it's used (and whether it's used often enough) before supporting it. If there is a strong sentiment that it should be supported out-of-the-gate, it'd be easy enough to check `CUDA_HOME` path in addition to `/usr/local/cuda*`. The main question is, when both `/usr/local/cuda` and `CUDA_HOME` are present and different (and both satisfy the specified version requirements), which one should be considered the default?

Hmm. For what it's worth, a platform I have access to sets all three of `CUDA_PATH`, `CUDA_HOME` and `CUDA_ROOT` to the same thing. I suppose that plays it safe.

### Proposition

I propose the following:

 1. If `CUDA_PATH` is defined and meets target/version requirements (if any), use that.
 2. If `CUDA_HOME` is defined and meets target/version requirements (if any), use that.
 3. If `CUDA_ROOT` is defined and meets target/version requirements (if any), use that.
 4. If there is a version requirement `x.y.z`,
     4.1. If `/usr/local/cuda-x.y` exists and meets target/version requirements (if any), use that.
     4.2. If `/usr/local/cuda` exists and meets target/version requirements (if any), use that.
5. Otherwise,
     5.1. If `/usr/local/cuda` exists and meets target requirements (if any), use that.
6. Abort.

### Rationale

- A user may not have administrator privileges. Therefore, Meson should not prefer privileged filesystem locations to environment variables, which are under control of the user.
- Environment variables are explicit statements by the user of a specific CUDA Toolkit location. Their existence should be considered strong evidence of a user preference.
- The `/usr/local/cuda` symlink is an explicit statement by the sysadmin or the package manager.
- `CUDA_PATH` is the closest thing to a standard, and is actually referenced in NVIDIA documentation. It can be set or changed by the user. It should have priority.
- The second most common variable (at least by Google hits and personal experience) is `CUDA_HOME`. If `CUDA_PATH` isn't defined but `CUDA_HOME` is, it's always acceptable to use it. A disagreement between `CUDA_PATH` and `CUDA_HOME` is the problem of the user; I would consider diagnosing it as a warning, but nothing more.
- `CUDA_ROOT` is even less common, but otherwise the same reasoning applies as for `CUDA_HOME`.
- Absent any explicit indications by the user of _where_ the CUDA Toolkit is, we can rely on the sysadmin "recommendation" and the `/usr/local` convention. If the user knows what version `x.y.z` of the Toolkit they want, trying `/usr/local/cuda-x.y` first seems like "the right thing to do" (more specific).
- The symlink `/usr/local/cuda` is less specific, and is something of a "last gasp", which is why it should be checked last, but it very often will exist. Most users who install just one CUDA Toolkit, or who make no effort to select something more specific, will be satisfied with `/usr/local/cuda`.

**Comment by agurtovoy** [comment]
@obilaniu Just FYI, rather than relying on the directory names, I'm actually reading `version.txt` from the root of each CUDA Toolkit dir to determine the version. I'll take a stab at tweaking my search algorithm to support `CUDA_HOME`/`CUDA_ROOT` in the way that you suggest and report back.

File: mesonbuild/dependencies/cuda.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        if defaults:
+            return (*defaults[0], True)
+
+        platform_msg = 'set the CUDA_PATH environment variable' if self._is_windows() \
+            else 'set the CUDA_PATH environment variable/create the \'/usr/local/cuda\' symbolic link'
+        msg = 'Please specify the desired CUDA Toolkit version (e.g. dependency(\'cuda\', version : \'>=10.1\')) or {} to point to the location of your desired version.'.format(platform_msg)
+        return self._report_dependency_error(msg, (None, None, False))
+
+    def _find_matching_toolkit(self, paths, version_reqs, nvcc_version):
+        # keep the default paths order intact, sort the rest in the descending order
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/cuda.py]: __[Flake8]__ trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/6076#issue-4738330))

<sub>posted by Sider</sub>

**Comment by agurtovoy** [comment]
Updated to support `CUDA_PATH`, `CUDA_HOME` and `CUDA_ROOT` on non-Windows OSes (`CUDA_PATH` is the standard on Windows and was already supported), added diagnostics for various possible misconfigurations, modified the toolkit selection logic to prefer the default location (if any) if it satisfies the version requirements, ready for review/merge.

**Comment by jpakkane** [comment]
Featurewise this is nice, haven't had time to review the content properly yet, though...

**Comment by agurtovoy** [comment]
@jpakkane Let me know if you have any questions.

**Comment by obilaniu** [comment]
@agurtovoy @jpakkane I have executed the testcases on my machine. All of them pass flawlessly, with a single hiccup: Currently, on my machine and with `10\ cuda\ dependency/version_reqs/meson.build`, the testcase fails because I only have CUDA Toolkits 10.0 and 10.1 installed, and a version req of `<10` will exclude both. If I switch it out to `<10.1`, then it will find my 10.0 install successfully, otherwise by default the module finds 10.1, as it indeed should.

I would personally recommend deleting that sub-testcase (because installing one CUDA Toolkit is pretty heavy, never mind two!) or else to make the testcase more robust use the constraint

    '!='+oldlib.version()

If that works? I know `>`, `>=`, `<`, `<=`, `==` work but don't know if `!=` does too.

File: mesonbuild/dependencies/cuda.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
+# Copyright 2013-2019 The Meson development team
```
Comment (jpakkane) [inline mesonbuild/dependencies/cuda.py]: Since this is an all new file, its copyright year should start from this year.

File: test cases/cuda/10 cuda dependency/c/meson.build [hunk 3]
```diff
@@ -0,0 +1 @@
+exe = executable('prog', 'prog.c', dependencies: dependency('cuda'))
```
Comment (jpakkane) [inline test cases/cuda/10 cuda dependency/c/meson.build]: There should be a `test()` that runs this executable to ensure that the binaries not only compile, but also work correctly when run. The same goes for all other test executables.

**Comment by jpakkane** [comment]
I tried this on an Ubuntu machine with Cuda 9 and a symlink from `/usr/local/cuda` to `/usr/local/cuda-9,0`. It fails because the SDK version does not match the nvcc version, which is 9.1.85. Using nvcc directly does work, though.

**Comment by agurtovoy** [comment]
@jpakkane Could you elaborate on how you arrived at this configuration? While in practice in most cases it's certainly possible to use a newer nvcc to compile against an older toolkit version, AFAIK there are no official guarantees on this. If this is common enough in the wild, we can always relax the check, but I'd love to learn more about how people get mismatched toolkit/compiler versions.

**Comment by jpakkane** [comment]
AFAICR I installed default Cuda packages either from the nonfree repo or from NVidia upstream.

**Comment by agurtovoy** [comment]
That is strange. Could you double-check where does your default `nvcc` come from in terms of filesystem location, and what does `/usr/local/cuda/bin/nvcc --version` say?

**Comment by agurtovoy** [comment]
Thanks for testing it out, @obilaniu! I'm currently leaning towards adding a check to the test requiring toolkit version < 10 to skip it if there are no matching toolkits, but I'm going to sleep on it. Ideally, this needs to be testing through unit tests, maybe that's what I'll do instead.

File: test cases/cuda/10 cuda dependency/c/meson.build (hunk 3 above)
Comment (agurtovoy) [reply test cases/cuda/10 cuda dependency/c/meson.build]: Done.

**Comment by agurtovoy** [comment]
Skipping the version requirements tests now if no matching toolkits were found, added a fallback version extraction from toolkit path if `version.txt` was not found (apparently if you install CUDA w/o docs on Windows, `version.txt` won't be there), tweaked the matching version search to ignore the patch version (which we can't detect if we're deducting version from path).

**Comment by agurtovoy** [comment]
@jpakkane Let me know when get a chance to check on https://github.com/mesonbuild/meson/pull/6076#issuecomment-548147105. FWIW, I just installed 9.0 from https://developer.nvidia.com/compute/cuda/9.0/Prod/local_installers/cuda_9.0.176_384.81_linux-run, and `/usr/local/cuda-9.0/bin/nvcc --version` gives me this:
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```

**Comment by jpakkane** [comment]
`dpkg -s` says that `nvcc` was installed by the `nvidia-cuda-toolkit` package. This is on a Ubuntu Bionic installation.

**Comment by agurtovoy** [comment]
@jpakkane Could you post the output of `cat /usr/local/cuda/version.txt` and `/usr/local/cuda/bin/nvcc --version`? Thank so much.

**Comment by jpakkane** [comment]
CUDA Version 9.0.176

and

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```

But plain `nvcc --version` gives

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Nov__3_21:07:56_CDT_2017
Cuda compilation tools, release 9.1, V9.1.85
```

So somehow I have two different Cuda toolkits installed?

**Comment by agurtovoy** [comment]
Yep, and the 9.1 bin path is (earlier) in `PATH`. My vote would be to keep the diagnostics that prevented you from compiling in this configuration, since it is somewhat sketchy; if we allowed meson to go ahead, the nvcc would be compiling against the 9.0 headers, but linking against the 9.1 runtime.

**Comment by jpakkane** [comment]
Yep, seems like a sensible thing to do. LGTM, anyone have comments still?

**Comment by obilaniu** [comment]
@jpakkane It works for me.

File: mesonbuild/dependencies/cuda.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        HasNativeKwarg.__init__(self, kwargs) # initialize self.for_machine
+        compilers = environment.coredata.compilers[self.for_machine]
+        language = self._detect_language(compilers)
+        if language not in self.supported_languages:
+            raise DependencyException('Language \'{}\' is not supported by the CUDA Toolkit. Supported languages are {}.'.format(language, self.supported_languages))
+
+        super().__init__('cuda', environment, language, kwargs)
+        self.requested_modules = self.get_requested(kwargs)
+        if 'cudart' not in self.requested_modules:
+            self.requested_modules = ['cudart'] + self.requested_modules
```
Comment (eli-schwartz) [inline mesonbuild/dependencies/cuda.py:40]: In https://github.com/elFarto/nvidia-vaapi-driver/issues/3#issuecomment-1005488544 there is apparently an objection to including libcudart in a use case where only libcuda is desired.

Since I don't know anything about the topic... Why is this necessary given you can already add it to the requested_modules via a kwarg?

Comment (agurtovoy) [reply mesonbuild/dependencies/cuda.py:40]: @eli-schwartz IIRC, I wanted the mainstream CUDA use case — which includes linking against the runtime (`cudart`) — to work out-of-the-box w/ as little boilerplate code on the user side as possible. That said, it looks like the way I did it does not allow one to override the default behavior, which is for sure a design bug. Please feel free to fix/revise.

Comment (eli-schwartz) [reply mesonbuild/dependencies/cuda.py:40]: @elFarto thoughts?

Comment (obilaniu) [reply mesonbuild/dependencies/cuda.py:40]: I don't recommend this. NVCC has the flag [`--cudart {none,shared,static}`](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#file-and-path-specifications-cudart) (default: `static`) for just this purpose, and the reason why `none` is allowed is that an application that goes to the trouble of shipping or embedding CUBINs and using only the CUDA Driver API can avoid any interaction with the CUDA Runtime API and rely only on the driver to call kernels. That makes it CUDA Toolkit-independent, which is an important positive. Code can `static`-ally link `libcudart` to get most of these advantages without most of the pain, but that makes it unsuitable for linking with code compiled by other CUDA Toolkits.

Comment (elFarto) [reply mesonbuild/dependencies/cuda.py:40]: I agree mostly with the way @agurtovoy designed it, as it's the most common use case. However, hard coding the cudart module in there is not ideal. Maybe we could add something like (I don't know Python too well):
```
if 'cudart' not in self.requested_modules and self.requested_modules != ['cuda']:
```
This should keep the default behaviour, and still allow overriding when you only want the driver, not the runtime. I think that you'd either want just the driver, or you'd want the runtime + maybe some other modules.

Comment (eli-schwartz) [reply mesonbuild/dependencies/cuda.py:40]: If I was designing this dependency today, I would definitely say "adding the module isn't really boilerplate IMO, so people should specify what they want and be allowed to not specify what they don't want".

The question is what to do about existing users... I'll defer to people who have a stake in the matter (and understand the stakes involved) to figure out the best way to detect the right thing to do. :)
